vsphere_plugin_common/__init__.py

```diff
@@ -154,6 +155,91 @@ def connect(self, cfg):
+        # in place to allow a clear path to the next release for users
+        allow_insecure = cfg.get('allow_insecure', False)
+
+        if certificate_path and allow_insecure:
+            raise NonRecoverableError(
+                'Cannot connect when certificate_path and allow_insecure '
+                'are both set. Unable to determine whether connection should '
+                'be secure or insecure.'
+            )
+        elif certificate_path:
+            if not hasattr(ssl, '_create_default_https_context'):
+                raise NonRecoverableError(
+                    'Cannot create secure connection with this version of '
+                    'python. This functionality requires at least python '
+                    '2.7.9 and has been confirmed to work on at least 2.7.12.'
+                )
+
+            if not os.path.exists(certificate_path):
+                raise NonRecoverableError(
+                    'Certificate was not found in {path}'.format(
+                        path=certificate_path,
+                    )
+                )
+            elif not os.path.isfile(certificate_path):
+                raise NonRecoverableError(
+                    'Found directory at {path}, but the certificate_path '
+                    'must be a file.'.format(
+                        path=certificate_path,
+                    )
+                )
+            try:
+                # We want to load the cert into the existing default context
+                # in case any other python modules have already defined their
+                # default https context.
+                ssl_context = ssl._create_default_https_context()
+                if ssl_context.verify_mode == 0:
+                    raise NonRecoverableError(
+                        'Default SSL context is not set to verify. '
+                        'Cannot use a certificate while other imported '
+                        'modules are disabling verification on the default '
+                        'SSL context.'
+                    )
+                ssl_context.load_verify_locations(certificate_path)
+            except ssl.SSLError as err:
+                if 'unknown error' in str(err).lower():
+                    raise NonRecoverableError(
+                        'Could not create SSL context with provided '
+                        'certificate {path}. This problem may be caused by '
+                        'the certificate not being in the correct format '
+                        '(PEM).'.format(
+                            host=host,
+                            path=certificate_path,
+                        )
+                    )
+                else:
+                    raise
+            def get_ssl_context(*args, **kwargs):
+                return ssl_context
+            ssl._create_default_https_context = get_ssl_context
+        else:
+            if allow_insecure:
```

Comment:
That'd need to be semi-flattened to:
elif allow_insecure and hasattr(ssl, '_creat_etc'):
        # warn, etc
elif allow_insecure:
        # warn, etc
else:
       # warn, etc

I think the logic flows more nicely the current way?